include/rsp.inc

```diff
@@ -0,0 +1,680 @@
+makeOpInstruction vrsql, 0b110101
+/** @brief Vector Accumulator Read (and Write) */
+makeOpInstruction vsar, 0b011101
+/** @brief Vector Subtraction of Short Elements */
+makeOpInstruction vsub, 0b010001
+/** @brief Vector Subtraction of Short Elements With Carry */
+makeOpInstruction vsubc, 0b010101
+/** @brief Vector XOR of Short Elements */
+makeOpInstruction vxor, 0b101100
+
+/** @brief Load Byte into Vector Register */
+makeLoadInstruction lbv, 0b00000
+/** @brief Load Double into Vector Register */
+makeLoadInstruction ldv, 0b00011
+/** @brief Load Packed Fourth into Vector Register */
+makeLoadInstruction lfv, 0b01001
+/** @brief Load Packed Half into Vector Register */
+makeLoadInstruction lhv, 0b01000
+/** @brief Load Long into Vector Register */
+makeLoadInstruction llv, 0b00010
+/** @brief Load Packed Bytes into Vector Register */
+makeLoadInstruction lpv, 0b00110
+/** @brief Load Quad into Vector Register */
+makeLoadInstruction lqv, 0b00100
+/** @brief Load Quad (Rest) into Vector Register */
+makeLoadInstruction lrv, 0b00101
+/** @brief Load Short into Vector Register */
+makeLoadInstruction lsv, 0b00001
+/** @brief Load Transpose into Vector Register */
+makeLoadInstruction ltv, 0b01011
+/** @brief Load Unsigned Packed into Vector Register */
+makeLoadInstruction luv, 0b00111
+
+/** @brief Store Byte from Vector Register */
+makeStoreInstruction sbv, 0b00000
+/** @brief Store Double from Vector Register */
+makeStoreInstruction sdv, 0b00011
+/** @brief Store Packed Fourth from Vector Register */
+makeStoreInstruction sfv, 0b01001
+/** @brief Store Packed Half from Vector Register */
+makeStoreInstruction shv, 0b01000
+/** @brief Store Long from Vector Register */
+makeStoreInstruction slv, 0b00010
+/** @brief Store Packed Bytes from Vector Register */
+makeStoreInstruction spv, 0b00110
+/** @brief Store Quad from Vector Register */
+makeStoreInstruction sqv, 0b00100
+/** @brief Store Quad (Rest) from Vector Register */
+makeStoreInstruction srv, 0b00101
+/** @brief Store Short from Vector Register */
+makeStoreInstruction ssv, 0b00001
+/** @brief Store Transpose from Vector Register */
+makeStoreInstruction stv, 0b01011
+/** @brief Store Unsigned Packed from Vector Register */
+makeStoreInstruction suv, 0b00111
+/** @brief Store Wrapped vector from Vector Register */
+makeStoreInstruction swv, 0b00111
+
+.macro mtc2 reg, vreg, element
+    hexRegisters
+    hexGeneralRegisters
+    .long (0x12 << 26 | 0x4 << 21 | \vreg << 11 | hex.\reg << 16 | \element << 7)
+.endm
+
+.macro mfc2 reg, vreg, element
+    hexRegisters
+    hexGeneralRegisters
+    .long (0x12 << 26 | 0x0 << 21 | \vreg << 11 | hex.\reg << 16 | \element << 7)
+.endm
+
+
+##################################################
+# Vector shift pseudo-opcodes
+##################################################
+# Helper macros for vector shifts.
+# How to use:
+#   1) Call macro "vsll_data" and/or "vsll8_data" in the data segment to define
```

Comment:
Can't we structure this to choose between the two such that `vsll <VDST>, <VSRC>, <QTY < 8>` will do `vsll_data` + `setup_vsll` + `vsll` and `vsll <VDST>, <VSRC>, <QTY >= 8>` do `vsll8_data` + `setup_vsll8` + `vsll8`?
Usage could be simplified to `vsll <VDST>, <VSRC>, <QTY>, <VTEMP>`

Reply:
We can do that because we need the setup part to be done once (possibly, once per microcode), allocating a global vector register. Of course there might be cases where you just need to shift once and in that case you could probably do it with a do-everything macro, but in the general case it's better to separate the setup from the actual shift instruction.

Reply:
Hmm I wonder if there is a way do something only once, it sounds possible. Regardless we can have a wrapper around it if we need that. Thx for the clarification.